persistence/sql/persister_identity.go

```diff
@@ -472,3 +510,115 @@ func (p *Persister) injectTraitsSchemaURL(ctx context.Context, i *identity.Ident
 	i.SchemaURL = s.SchemaURL(p.r.Config(ctx).SelfPublicURL(nil)).String()
 	return nil
 }
+
+func (p *Persister) getJsonSearchQuery(field string, values []string) pop.ScopeFunc {
+	return func(q *pop.Query) *pop.Query {
+		switch p.Connection(context.Background()).Dialect.Name() {
+		case "sqlite3", "mysql", "mariadb":
+			field, innerField := extractFieldAndInnerFields(field)
+			for _, value := range values {
+				if innerField == "" {
+					q = q.Where(fmt.Sprintf(`json_extract(%s, '$') = ?`, field), value)
+				} else {
+					stmt := fmt.Sprintf(`json_extract(%s, '$.%s') = ?`, field, innerField)
+					q = q.Where(stmt, value)
+				}
+			}
+			return q
+		case "postgres", "cockroach":
+			field, innerField := extractFieldAndInnerFields(field)
+			for _, value := range values {
+				if innerField == "" {
+					q = q.Where(fmt.Sprintf(`%s @> '"%s"'`, field, value))
+				} else {
+					q = q.Where(fmt.Sprintf(`%s @> '{"%s":"%s"}'`, field, innerField, value))
+				}
+			}
+			return q
+		}
+		return q
+	}
+}
+
+func (p *Persister) searchCredentialQuery(field string, values []string) pop.ScopeFunc {
+	var innerField string
+	_, innerField = extractFieldAndInnerFields(field)
+	switch innerField {
+	case "type":
+		return func(q *pop.Query) *pop.Query { return q.Where("credential_types.name IN (?)", values) }
+	case "identifier":
+		return func(q *pop.Query) *pop.Query { return q.Where("credential_identifiers.identifier IN (?)", values) }
+	default:
+		return func(q *pop.Query) *pop.Query { return q }
+	}
+}
+
+var quoteChar = map[string]uint8{
+	"cockroach": '"',
+	"mariadb":   '`',
+	"mysql":     '`',
+	"postgres":  '"',
+	"sqlite3":   '"',
+}
+
+func (p *Persister) Quote(key string, c uint8) string {
+	if c == 0 {
+		c = '"'
+	}
+	parts := strings.Split(key, ".")
+
+	for i, part := range parts {
+		part = strings.Trim(part, `"`)
+		part = strings.Trim(part, "`")
+		part = strings.TrimSpace(part)
+
+		parts[i] = fmt.Sprintf(`%c%v%c`, c, part, c)
+	}
```

Comment:
What happens if `key` is `foo"; DROP TABLES ALL; --` or something similar (assuming quote `"`)?

Reply:
I would also change this func to do the quote detection itself, as it is already defined on the persister.
```suggestion
func (p *Persister) Quote(ctx context.Context, key string) string {
	n := p.Connection(ctx).Dialect.Name()
	c, ok := quoteChar[n]
	if !ok {
		// guess panic is OK here as the error is not fixable without a new release of Kratos
		panic("DSN is of unknown dialect " + n)
	}

	parts := strings.Split(key, ".")

	for i, part := range parts {
		part = strings.Trim(part, `"`)
		part = strings.Trim(part, "`")
		part = strings.TrimSpace(part)

		parts[i] = fmt.Sprintf(`%c%v%c`, c, part, c)
	}
```

Reply:
> What happens if `key` is `foo"; DROP TABLES ALL; --` or something similar (assuming quote `"`)?

does not seem to work